packages/react/src/runtime/style-cache.tsx

```diff
@@ -1,24 +1,70 @@
 import * as React from 'react';
 import { createContext, useContext } from 'react';
 import { isNodeEnvironment } from './is-node';
+import insertRule from './sheet';
 import { ProviderComponent, UseCacheHook } from './types';
 
 /**
  * Cache to hold already used styles.
  * React Context on the server - singleton object on the client.
  */
-const Cache: any = isNodeEnvironment() ? createContext<Record<string, true> | null>(null) : {};
+let Cache: any = isNodeEnvironment() ? createContext<Record<string, true> | null>(null) : {};
 
 if (!isNodeEnvironment()) {
+  const ssrCache = document.querySelector('script[data-cmpld="c"]');
+  if (ssrCache) {
+    try {
+      Cache = { ...Cache, ...JSON.parse(ssrCache.innerHTML) };
+      ssrCache.remove();
+    } catch (e) {
+      console.log('Error reading SSR cache', e);
+    }
+  }
+
   /**
    * Iterates through all found style elements generated when server side rendering.
-   *
-   * @param cb
    */
-  const ssrStyles = document.querySelectorAll<HTMLStyleElement>('style[data-cmpld]');
+  const ssrStyles = document.querySelectorAll<HTMLStyleElement>('style[data-cmpld="s"]');
   for (let i = 0; i < ssrStyles.length; i++) {
-    // Move all found server-side rendered style elements to the head before React hydration happens.
-    document.head.appendChild(ssrStyles[i]);
+    // Create the client cache for all SSR'd classes
+    const stylesheet = ssrStyles[i];
+    // move all SSR style tags to the head straight away
+    document.head.appendChild(stylesheet);
```

Comment:
We move these to the head so the styles don't disappear. Then we go through and consolidate them.